### PR TITLE
Add tensorflow-keras integration

### PR DESCRIPTION
**Pull request by HubertJaworski**

### CODE REVIEW

**Comment by HubertJaworski** [comment]
Needs review toghether with https://github.com/neptune-ai/neptune-contrib/pull/126

File: integrations/tensorflow-keras/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * [neptune-contrib readthedocs](https://neptune-contrib.readthedocs.io)
+ * [neptune community forum](https://community.neptune.ai/)
+ * [neptune community Spectrum](https://spectrum.chat/neptune-community?tab=posts)
+ * Github issues
+ 
+# Contributing
+If you see something that you don't like you are more than welcome to contribute!
+There are many options:
+  
+  * Participate in discussions on [neptune community forum](https://community.neptune.ai/) or [neptune community Spectrum](https://spectrum.chat/neptune-community?tab=posts)
```
Comment (aniezurawski) [inline integrations/tensorflow-keras/README.md:45]: Ask Mycek what should be here instead.

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    except ImportError:
+        msg = """
+        keras package not found. 
+
+        As Keras is now part of Tensorflow you should install it by running
+            pip install tensorflow"""
+        raise ModuleNotFoundError(msg) # pylint:disable=undefined-variable
+
+
+class NeptuneMonitor(Callback):
```
Comment (aniezurawski) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:38]: NeptuneCallback according to spec.

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    """Logs Keras metrics to Neptune.
+
+    Goes over the `last_metrics` and `smooth_loss` after each batch and epoch
+    and logs them to Neptune.
+
+    If experiment parameter is not provided, Neptune will attempt to self-initialize using
+    `project` and  `api_token` parameters
```
Comment (aniezurawski) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:45]: I cannot see it in spec.

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            model.fit(x_train, y_train,
+                      epochs=PARAMS['epoch_nr'],
+                      batch_size=PARAMS['batch_size'],
+                      callbacks=[NeptuneMonitor()])
+
+    Note:
+        You need to have Keras or Tensorflow 2 installed on your computer to use this module.
+    """
+
+    def __init__(self, experiment=None, prefix='', project=None, api_token=None):
```
Comment (aniezurawski) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:97]: `def __init__(self, experiment, base_namespace=””)` according to spec.

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
+#
```
Comment (aniezurawski) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:1]: @Herudaio could you review this file as well?

File: integrations/tensorflow-keras/setup.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    all_deps = []
+    for group_name in extras:
+        all_deps += extras[group_name]
+    extras['all'] = all_deps
+
+    base_libs = ['neptune-client>=0.4.126']
+
+    setup(
+        name='neptune-tensorflow-keras',
+        version="0.0.1", # TODO fix version generation, propably use versioneer
```
Comment (aniezurawski) [inline integrations/tensorflow-keras/setup.py:20]: Can we release different pip packages in this repo separately? Should we? How would it work?

**Review by aniezurawski — CHANGES_REQUESTED** [review]

File: neptune/alpha/integrations/tensorflow_keras/__init__.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+try:
+    from neptune.alpha.integrations.tensorflow_keras.impl import NeptuneMonitor
```
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py:18]: user will use it like this

File: integrations/tensorflow-keras/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# neptune-contrib
```
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply integrations/tensorflow-keras/README.md:1]: copy paste, I'll empty it and we'll write it again

File: neptune/alpha/integrations/tensorflow_keras/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aniezurawski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py:18]: No. User will import from this module. User will never import anything from neptune-tensorflow-keras package directly.

File: neptune/alpha/integrations/tensorflow_keras/__init__.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    NeptuneMonitor = NeptuneMonitor
```
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py]: already removed

File: neptune/alpha/exceptions.py [hunk 10]
```diff
@@ -340,3 +340,24 @@ def __init__(self):
 {correct}Need help?{end}-> TODO: create and link docs
 """.format(**STYLES)
         super().__init__(message)
+
+
+class NeptuneIntegrationNotInstalledException(NeptuneException):
+    def __init__(self, neptune_framwork, framwork):
```
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/alpha/exceptions.py:346]: We can, but this would lock us to naming convention (neptune-<some_framework> for package name and <some_framework> for extras

[the remark this answers is not in the document]
Comment (aniezurawski) [reply neptune/alpha/exceptions.py:346]: Is it wrong? Sounds like it should be this way.

File: integrations/tensorflow-keras/setup.py (hunk 6 above)
Comment (HubertJaworski) [reply integrations/tensorflow-keras/setup.py:20]: Lets make this a separe MR - I suggest using different tag regexps - for example `tensorflow-keras/a.b.c` would release only this integration while `a.b.c` would release main package

Comment (aniezurawski) [reply integrations/tensorflow-keras/setup.py:20]: Ok.

File: neptune/alpha/integrations/tensorflow_keras/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py]: To reiterate:
The only oficially supported way to import integration is `from neptune.alpha.integrations.tensorflow_keras.impl import NeptuneMonitor`.
User should not import it in any other way like  `from neptune_tensorflow_keras import NeptuneMonitor` or from within `impl` module

[the remark this answers is not in the document]
Comment (PiotrJander) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py]: But we should not expose the `impl` package to the user

From mycek's spec: `from neptune.new.integrations.tensorflow_keras import NeptuneCallback`

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+            experiment = neptune.init(api_token='ANONYMOUS',
+                                      project='shared/keras-integration')
+
+        Instantiate the monitor and pass
+        it to callbacks argument of `model.fit()`:
+
+        .. code:: python
+
+            from neptunecontrib.monitoring.keras import NeptuneMonitor
```
Comment (PiotrJander) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:72]: The docstring is outdated, we should remove or update it

File: integrations/tensorflow-keras/setup.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    extras = {
+    }
+
+    all_deps = []
+    for group_name in extras:
+        all_deps += extras[group_name]
+    extras['all'] = all_deps
+
+    base_libs = ['neptune-client>=0.5.1']
```
Comment (PiotrJander) [inline integrations/tensorflow-keras/setup.py:16]: I don't think this will be needed in this package - get rid of those lines?

File: neptune/alpha/exceptions.py [hunk 13]
```diff
@@ -340,3 +340,24 @@ def __init__(self):
 {correct}Need help?{end}-> TODO: create and link docs
 """.format(**STYLES)
         super().__init__(message)
+
+
+class NeptuneIntegrationNotInstalledException(NeptuneException):
+    def __init__(self, framwork):
+        message = """
```
Comment (PiotrJander) [inline neptune/alpha/exceptions.py:347]: Perhaps make this local variable into a top-level variable? Multi-line strings look ugly in the middle of the method

**Review by PiotrJander — CHANGES_REQUESTED** [review]

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        if not logs:
+            return
+
+        prefix = self._namespace + trigger + '/'
+
+        for metric, value in logs.items():
+            try:
+                if metric in ('batch', 'size'):
+                    continue
+                name = prefix + metric
```
Comment (PiotrJander) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:105]: We shouldn't do string concatenation, do `e[namespace][trigger][metric]` instead.

Or

```
ns = e[namespace]
ns[trigger][metric].log(...)
```

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        You need to have Keras or Tensorflow 2 installed on your computer to use this module.
+    """
+
+    def __init__(self, experiment: Experiment, base_namespace: Optional[str] = None):
+        super().__init__()
+        if experiment is None or not isinstance(experiment, Experiment):
+            raise ValueError("Neptune experiment is missing")
+        self._exp = experiment
+        self._namespace = ''
+        if base_namespace:
```
Comment (PiotrJander) [inline integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py:89]: We shouldn't do string concatenation, do `e[namespace][trigger][metric]` instead.

Or

```
ns = e[namespace]
ns[trigger][metric].log(...)
```

This way checking if the string ends with `/` is unnecessary.

**Review by PiotrJander — CHANGES_REQUESTED** [review]

File: neptune/alpha/integrations/tensorflow_keras/__init__.py [hunk 16]
```diff
@@ -15,7 +15,7 @@
 #
 
 try:
-    from neptune.alpha.integrations.tensorflow_keras.impl import NeptuneCallback
+    from neptune_tensorflow_keras.impl import NeptuneCallback
```
Comment (PiotrJander) [inline neptune/alpha/integrations/tensorflow_keras/__init__.py:18]: Ten `impl` już tu chyba nie potrzebny?

Comment (HubertJaworski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py:18]: good catch

File: integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py (hunk 15 above)
Comment (HubertJaworski) [reply integrations/tensorflow-keras/neptune/alpha/integrations/tensorflow_keras/impl/__init__.py]: this works?

File: integrations/tensorflow-keras/setup.py (hunk 12 above)
Comment (HubertJaworski) [reply integrations/tensorflow-keras/setup.py:17]: we will change this to minimal supported client version after i gets released

File: neptune/alpha/exceptions.py (hunk 13 above)
Comment (HubertJaworski) [reply neptune/alpha/exceptions.py:347]: This is done in the exact same way as the rest of this file. Making it top-level doesn't make sense

**Review by PiotrJander — APPROVED** [review]

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+
+class NeptuneCallback(Callback):
+    """Logs Keras metrics to Neptune.
+
+    Goes over the `last_metrics` and `smooth_loss` after each batch and epoch
+    and logs them to Neptune.
+
+    If experiment parameter is not provided, Neptune will attempt to self-initialize using
+    `project` and  `api_token` parameters
```
Comment (Herudaio) [inline integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:51]: To remove. Experiment parameter is now required.

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+
+    See the example experiment here https://ui.neptune.ai/shared/keras-integration/e/KERAS-23/logs
+
+    Args:
+        experiment: `neptune.Experiment`:
```
Comment (Herudaio) [inline integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:56]: `neptune.new.Experiment`?

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+        base_namespace: str, optional:
+            Namespace, in which all series will be put.
+
+    Example:
+
+        Initialize Neptune client:
+
+        .. code:: python
+
+            import neptune
```
Comment (Herudaio) [inline integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:67]: `import neptune.new as neptune`

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+
+            experiment = neptune.init(api_token='ANONYMOUS',
+                                      project='shared/keras-integration')
```
Comment (Herudaio) [inline integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:70]: Project name should be updated. @kamil-kaczmarek will know new proper name.

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+
+        Instantiate the monitor and pass
+        it to callbacks argument of `model.fit()`:
+
+        .. code:: python
+
+            from neptune.alpha.integrations.tensorflow_keras import NeptuneCallback
```
Comment (Herudaio) [inline integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:77]: `neptune.alpha` -> `neptune.new`

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+        Initialize Neptune client:
+
+        .. code:: python
+
+            import neptune
+
+            experiment = neptune.init(api_token='ANONYMOUS',
+                                      project='shared/keras-integration')
+
+        Instantiate the monitor and pass
```
Comment (Herudaio) [inline integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:72]: `the monitor` -> `the callback`

File: neptune/alpha/exceptions.py [hunk 23]
```diff
@@ -340,3 +340,24 @@ def __init__(self):
 {correct}Need help?{end}-> TODO: create and link docs
 """.format(**STYLES)
         super().__init__(message)
+
+
+class NeptuneIntegrationNotInstalledException(NeptuneException):
+    def __init__(self, framwork):
```
Comment (Herudaio) [inline neptune/alpha/exceptions.py:346]: `framwork` -> `framework` :)

**Review by Herudaio — CHANGES_REQUESTED** [review]

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py (hunk 18 above)
Comment (HubertJaworski) [reply integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:56]: we will do a single, bulk edit alpha -> new in future. Utlill it's done, it should be alpha

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py (hunk 20 above)
Comment (HubertJaworski) [reply integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:70]: It doesn't exist yet, it should be in `common` organization. we will update it once it gets set up

File: integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py (hunk 21 above)
Comment (HubertJaworski) [reply integrations/tensorflow-keras/neptune_tensorflow_keras/__init__.py:77]: not now

File: neptune/alpha/integrations/tensorflow_keras/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py]: guys, it was `from neptune.new.integrations.tensorflow_keras import NeptuneCallback` from the beginning :P

**Review by Herudaio — APPROVED** [review]

File: neptune/alpha/integrations/tensorflow_keras/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aniezurawski) [reply neptune/alpha/integrations/tensorflow_keras/__init__.py]: Mycek spec is about something else. It's about this commented file, not imported one. xD
